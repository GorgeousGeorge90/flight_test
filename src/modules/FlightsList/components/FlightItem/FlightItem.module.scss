

.ticket_container {
  @apply flex items-center flex-col;

  .ticket_content {
    @apply flex flex-col items-center justify-center gap-5;

    .ticket_header {
      @apply flex items-center bg-blue-400 text-neutral-50 w-full px-5 gap-2;
      height: 50px;

      .ticket_price {
        @apply flex flex-col items-end flex-grow;
      }
    }

    .ticket_list {
      @apply flex flex-col gap-2;


      .ticket_main {
        @apply flex flex-col gap-1 p-2;

        .ticket_time {
          @apply flex gap-10 items-center justify-center font-mono;

          .ticket_rule {
            @apply text-orange-500;
          }
        }

        .ticket_duration {
          @apply flex flex-col items-center;
        }
      }
    }

    .ticket_footer {
      @apply flex items-center justify-center w-full bg-orange-400 py-2 text-neutral-100;

      button {
        @apply uppercase;
      }
    }
  }
}

