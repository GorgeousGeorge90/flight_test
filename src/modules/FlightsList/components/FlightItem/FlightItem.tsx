import { FlightType } from '../../types/types';
import styles from './FlightItem.module.scss';
import icon from './../../../../assets/img/icon.png';
import { CurrentType } from '../../FlightsList';



export type FlightItemProps = {
    flight: FlightType,
    setActive: (param:boolean) => void,
    setCurrent: (param:CurrentType) => void,
}


const FlightItem = ({flight,setActive,setCurrent}:FlightItemProps) => {

    return (<section className={styles.ticket_container}>
        <div className={styles.ticket_content}>
            <header className={styles.ticket_header}>
                <span><img src={icon} alt='icon'/></span>
                <p style={{fontWeight: 'bold', fontStyle: 'italic'}}>{flight.carrier}</p>
                <p className={styles.ticket_price}>
                    <span style={{fontSize: 'larger'}}>{flight.price} &#8381;</span>
                    <span style={{fontSize: 'x-small'}}>Cтоимость для одного взрослого пассажира</span>
                </p>
            </header>
            <ul className={styles.ticket_list}>
                {
                    flight.tickets.map((item, i) => <li key={i}>
                        <main className={styles.ticket_main}>
                            <p>
                                {item.ticket.departureCity},&nbsp;
                                <span>{item.ticket.departureAirport.caption}</span>&nbsp;
                                <span style={{color: 'blue'}}>( {item.ticket.departureAirport.uid} )</span>
                                &nbsp;&rarr;&nbsp;
                                {item.ticket.arrivalCity},&nbsp;
                                <span>{item.ticket.arrivalAirport.caption}</span>&nbsp;
                                <span style={{color: 'blue'}}>( {item.ticket.arrivalAirport.uid} )</span>
                            </p>
                            <hr/>
                            <span className={styles.ticket_time}>
                                <b>{item.ticket.departureDate}</b>&rarr;
                                <span className={styles.ticket_duration}>
                                <p>&#9201; {item.duration}</p>
                                    {
                                        item.ticket.hasOwnProperty('id') ? <p className={styles.ticket_rule}
                                                            onClick={() => {
                                                                const payload = {
                                                                    id:item.ticket.id,
                                                                    index:item.ticket.index,
                                                                }

                                                                setCurrent(payload as CurrentType)
                                                                setActive(true)
                                                            }}>1 пересадка</p>:null
                                    }
                            </span>&rarr;
                                <b>{item.ticket.arrivalDate}</b>
                    </span>
                            <hr/>
                            <p>
                                Рейс выполняет: {item.ticket.operatingAirline}
                            </p>
                            <hr style={{borderWidth: '2px', borderColor: 'darkgray'}}/>
                        </main>
                    </li>)
                }
            </ul>
            <footer className={styles.ticket_footer}>
                <button>выбрать</button>
            </footer>
        </div>
    </section>)
}

export default FlightItem