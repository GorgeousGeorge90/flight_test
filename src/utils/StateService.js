import SupportService from './SupportService';

class StateService {
    createState(data) {
        return data.flights.reduce((acc,flight,i) => {
            let item = {
                id: i,
                carrier:flight.flight.carrier.caption,
                price: flight.flight.price.total.amount,
                tickets: flight.flight.legs.reduce((acc, ticket) => {
                    let item = {
                        duration:SupportService.durationStyle(ticket.duration),
                        ticket: ticket.segments.reduce((acc, segment) => {
                            let item = {
                                departureAirport: segment.departureAirport,
                                departureCity: segment.departureCity.caption,
                                arrivalDate: SupportService.dateTransform(segment.arrivalDate),
                                departureDate: SupportService.dateTransform(segment.departureDate),
                                arrivalCity: segment.arrivalCity.caption,
                                arrivalAirport: segment.arrivalAirport,
                                airline: segment.airline.caption,
                                operatingAirline: segment.operatingAirline ? segment.operatingAirline.caption:segment.airline.caption,
                                travelDuration:segment.travelDuration,
                            }

                            acc.push(item)

                            return acc
                        }, [])
                    }

                    acc.push(item)

                    return acc
                }, [])
            }

            acc.push(item)
            return acc

        },[])
    }

    createExtraState(state) {
         const new_state = JSON.parse(JSON.stringify(state))
         return new_state.map(item => {
            return {
                ...item,
                tickets: item.tickets.reduce((acc,ticket,i) => {
                    if (ticket.ticket.length === 1) {
                        ticket.ticket = ticket.ticket[0]
                        acc.push(ticket)
                    } else {
                        ticket.ticket = {
                            id:item.id,
                            index:i,
                            departureAirport: ticket.ticket[0].departureAirport,
                            departureCity: ticket.ticket[0].departureCity,
                            departureDate: ticket.ticket[0].departureDate,
                            arrivalDate: ticket.ticket[1].arrivalDate,
                            arrivalCity: ticket.ticket[1].arrivalCity,
                            arrivalAirport: ticket.ticket[1].arrivalAirport,
                            operatingAirline: ticket.ticket[1].operatingAirline,
                        }
                        acc.push(ticket)
                    }

                    return acc
                },[])
            }
        })
    }

}

export default new StateService()