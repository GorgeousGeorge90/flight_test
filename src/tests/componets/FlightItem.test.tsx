import FlightItem, { FlightItemProps } from '../../modules/FlightsList/components/FlightItem/FlightItem';
import {render, screen} from '@testing-library/react';
import StateService from '../../utils/StateService';
import data from '../../data/data';


const state = StateService.createState(data)
const extra_state = StateService.createExtraState(state)

describe('', () => {
    it('render FlightItem component', () => {
            const data: FlightItemProps = {
                flight: extra_state[0],
                setActive: jest.fn(),
                setCurrent: jest.fn(),
            }
            const view = render(<FlightItem {...data}/>)
            expect(view).toMatchSnapshot()
    })
})
