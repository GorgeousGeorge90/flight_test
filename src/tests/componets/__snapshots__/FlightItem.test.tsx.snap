// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` render FlightItem component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="ticket_container"
      >
        <div
          class="ticket_content"
        >
          <header
            class="ticket_header"
          >
            <span>
              <img
                alt="icon"
                src="icon.png"
              />
            </span>
            <p
              style="font-weight: bold; font-style: italic;"
            >
              Air France
            </p>
            <p
              class="ticket_price"
            >
              <span
                style="font-size: larger;"
              >
                105368
                 ₽
              </span>
              <span
                style="font-size: x-small;"
              >
                Cтоимость для одного взрослого пассажира
              </span>
            </p>
          </header>
          <ul
            class="ticket_list"
          >
            <li>
              <main
                class="ticket_main"
              >
                <p>
                  Москва
                  , 
                  <span>
                    ШЕРЕМЕТЬЕВО
                  </span>
                   
                  <span
                    style="color: blue;"
                  >
                    ( 
                    SVO
                     )
                  </span>
                   → 
                  ЛОНДОН
                  , 
                  <span>
                    Лондон, Хитроу
                  </span>
                   
                  <span
                    style="color: blue;"
                  >
                    ( 
                    LHR
                     )
                  </span>
                </p>
                <hr />
                <span
                  class="ticket_time"
                >
                  <b>
                    6:10 18.7
                  </b>
                  →
                  <span
                    class="ticket_duration"
                  >
                    <p>
                      ⏱ 
                      12 ч 15 мин
                    </p>
                    <p
                      class="ticket_rule"
                    >
                      1 пересадка
                    </p>
                  </span>
                  →
                  <b>
                    16:25 18.7
                  </b>
                </span>
                <hr />
                <p>
                  Рейс выполняет: 
                  Air France
                </p>
                <hr
                  style="border-width: 2px; border-color: darkgray;"
                />
              </main>
            </li>
            <li>
              <main
                class="ticket_main"
              >
                <p>
                  ЛОНДОН
                  , 
                  <span>
                    Лондон, Хитроу
                  </span>
                   
                  <span
                    style="color: blue;"
                  >
                    ( 
                    LHR
                     )
                  </span>
                   → 
                  Москва
                  , 
                  <span>
                    ШЕРЕМЕТЬЕВО
                  </span>
                   
                  <span
                    style="color: blue;"
                  >
                    ( 
                    SVO
                     )
                  </span>
                </p>
                <hr />
                <span
                  class="ticket_time"
                >
                  <b>
                    17:35 19.7
                  </b>
                  →
                  <span
                    class="ticket_duration"
                  >
                    <p>
                      ⏱ 
                      23 ч 20 мин
                    </p>
                    <p
                      class="ticket_rule"
                    >
                      1 пересадка
                    </p>
                  </span>
                  →
                  <b>
                    18:55 20.7
                  </b>
                </span>
                <hr />
                <p>
                  Рейс выполняет: 
                  Аэрофлот - российские авиалинии
                </p>
                <hr
                  style="border-width: 2px; border-color: darkgray;"
                />
              </main>
            </li>
          </ul>
          <footer
            class="ticket_footer"
          >
            <button>
              выбрать
            </button>
          </footer>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="ticket_container"
    >
      <div
        class="ticket_content"
      >
        <header
          class="ticket_header"
        >
          <span>
            <img
              alt="icon"
              src="icon.png"
            />
          </span>
          <p
            style="font-weight: bold; font-style: italic;"
          >
            Air France
          </p>
          <p
            class="ticket_price"
          >
            <span
              style="font-size: larger;"
            >
              105368
               ₽
            </span>
            <span
              style="font-size: x-small;"
            >
              Cтоимость для одного взрослого пассажира
            </span>
          </p>
        </header>
        <ul
          class="ticket_list"
        >
          <li>
            <main
              class="ticket_main"
            >
              <p>
                Москва
                , 
                <span>
                  ШЕРЕМЕТЬЕВО
                </span>
                 
                <span
                  style="color: blue;"
                >
                  ( 
                  SVO
                   )
                </span>
                 → 
                ЛОНДОН
                , 
                <span>
                  Лондон, Хитроу
                </span>
                 
                <span
                  style="color: blue;"
                >
                  ( 
                  LHR
                   )
                </span>
              </p>
              <hr />
              <span
                class="ticket_time"
              >
                <b>
                  6:10 18.7
                </b>
                →
                <span
                  class="ticket_duration"
                >
                  <p>
                    ⏱ 
                    12 ч 15 мин
                  </p>
                  <p
                    class="ticket_rule"
                  >
                    1 пересадка
                  </p>
                </span>
                →
                <b>
                  16:25 18.7
                </b>
              </span>
              <hr />
              <p>
                Рейс выполняет: 
                Air France
              </p>
              <hr
                style="border-width: 2px; border-color: darkgray;"
              />
            </main>
          </li>
          <li>
            <main
              class="ticket_main"
            >
              <p>
                ЛОНДОН
                , 
                <span>
                  Лондон, Хитроу
                </span>
                 
                <span
                  style="color: blue;"
                >
                  ( 
                  LHR
                   )
                </span>
                 → 
                Москва
                , 
                <span>
                  ШЕРЕМЕТЬЕВО
                </span>
                 
                <span
                  style="color: blue;"
                >
                  ( 
                  SVO
                   )
                </span>
              </p>
              <hr />
              <span
                class="ticket_time"
              >
                <b>
                  17:35 19.7
                </b>
                →
                <span
                  class="ticket_duration"
                >
                  <p>
                    ⏱ 
                    23 ч 20 мин
                  </p>
                  <p
                    class="ticket_rule"
                  >
                    1 пересадка
                  </p>
                </span>
                →
                <b>
                  18:55 20.7
                </b>
              </span>
              <hr />
              <p>
                Рейс выполняет: 
                Аэрофлот - российские авиалинии
              </p>
              <hr
                style="border-width: 2px; border-color: darkgray;"
              />
            </main>
          </li>
        </ul>
        <footer
          class="ticket_footer"
        >
          <button>
            выбрать
          </button>
        </footer>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
